libraries/AC_PrecLand/AC_PrecLand_StateMachine.cpp

```diff
@@ -0,0 +1,254 @@
+#include "AC_PrecLand_StateMachine.h"
+
+static const float MAX_POS_ERROR_CM = 75.0f;  // Maximum possition error for retry locations
+static const uint32_t FAILSAFE_INIT_TIMEOUT_MS = 7000;   // Timeout in ms before failsafe measures are started. During this period vehicle is completely stopped to give user the time to take over
+static const float RETRY_OFFSET_ALT_CM = 150.0f;  // This gets added to the altitude of the retry location
+
+// Constructor
+AC_PrecLand_StateMachine::AC_PrecLand_StateMachine(const AP_InertialNav& inav) :
+    _inertial_nav(inav)
+{
+}
+
+// Initialize the state machine. This is called everytime vehicle switches mode
+void AC_PrecLand_StateMachine::init()
+{
+    // init is only called ONCE per mode change. So in a particuar mode we can retry only a finite times.
+    // The counter will be reset if the statemachine is called from a different mode
+    _retry_count = 0;
+    // reset every other statemachine
+    reset_failed_landing_statemachines();
+}
+
+// Reset the landing statemachines. This needs to be called everytime the landing target is back in sight.
+// So that if the landing target goes out of sight again, we can start the failed landing procedure back from the beginning stage
+void AC_PrecLand_StateMachine::reset_failed_landing_statemachines()
+{
+    landing_target_lost_action = TargetLostAction::INIT;
+    retry_state = RetryLanding::INIT;
+    failsafe_initialized = false;
+}
+
+// Run Prec Land State Machine. During Prec Landing, we might encounter four scenarios:
+// 1. We had the target in sight, but have lost it now. 2. We never had the target in sight and user wants to land.
+// 3. We have the target in sight and can continue landing. 4. The sensor is out of range
+// This method deals with all of these scenarios
+// Returns the action needed to be done by the vehicle.
+// Parameters: Vector3f "retry_pos_cm" is filled with the required location if we need to retry landing.
+AC_PrecLand_StateMachine::Status AC_PrecLand_StateMachine::update_precland_state_machine(Vector3f &retry_pos_cm)
+{
+    // This variable will get updated as the situation demands
+    Status precland_status{Status::DESCEND};
+
+    // grab the current status of Landing Target
+    AC_PrecLand *_precland = AP::ac_precland();
+    if (_precland == nullptr) {
+        // should never happen
+        return precland_status = Status::ERROR;
```

Comment:
you could raise an internal error here to make the problem more visible in SITL and to increase the chance that users will report the issue so we can resolve it.